tfs/frame.py

```diff
@@ -338,39 +341,39 @@ def _element_is_list(element):
-                LOGGER.debug(
-                    f"'None' values found in headers of dataframe {info_str}, which is incompatible with MAD-X."
-                    "Remove them or assign a value in order to keep compatibility with MAD-X."
-                )
-                errmsg = "TFS-Headers can not contain 'None' values in MAD-X compatibility mode."
-                raise MADXCompatibilityError(errmsg)
-
-            # MAD-X will not accept back in a TFS file with no 'TYPE' entry in the headers (as string)
-            if "TYPE" not in data_frame.headers:
-                LOGGER.warning("MAD-X expects a 'TYPE' header in the TFS file, which is missing. Adding it.")
-                data_frame.headers["TYPE"] = "Added by tfs-pandas for MAD-X compatibility"
+        
+        # Check that no boolean values are in the headers - MAD-X does not accept them
+        if _has_headers(data_frame) and any(isinstance(header, bool) for header in data_frame.headers.values()):
+            LOGGER.debug(
+                f"Boolean values found in headers of dataframe {info_str}, which is incompatible with MAD-X."
+                "Change their types in order to keep compatibility with MAD-X."
+            )
+            errmsg = "TFS-Headers can not contain boolean values in MAD-X compatibility mode."
+            raise MADXCompatibilityError(errmsg)
+
+        # Check that no complex values are in the headers - MAD-X does not accept them
+        if _has_headers(data_frame) and any(isinstance(header, complex) for header in data_frame.headers.values()):
+            LOGGER.debug(
+                f"Complex values found in headers of dataframe {info_str}, which is incompatible with MAD-X."
+                "Change their types in order to keep compatibility with MAD-X."
+            )
+            errmsg = "TFS-Headers can not contain complex values in MAD-X compatibility mode."
+            raise MADXCompatibilityError(errmsg)
+
+        # Check that no 'None' values are in the headers - it would
+        # write as 'nil' which MAD-X does not accept
+        if _has_headers(data_frame) and any(header is None for header in data_frame.headers.values()):
+            LOGGER.debug(
+                f"'None' values found in headers of dataframe {info_str}, which is incompatible with MAD-X."
+                "Remove them or assign a value in order to keep compatibility with MAD-X."
+            )
+            errmsg = "TFS-Headers can not contain 'None' values in MAD-X compatibility mode."
+            raise MADXCompatibilityError(errmsg)
+
+        # MAD-X will not accept back in a TFS file with no 'TYPE' entry in the headers (as string)
+        if _has_headers(data_frame) and "TYPE" not in data_frame.headers:
+            LOGGER.warning("MAD-X expects a 'TYPE' header in the TFS file, which is missing. Adding it.")
+            data_frame.headers["TYPE"] = "Added by tfs-pandas for MAD-X compatibility"
```

Comment:
This is not good. Not a valid dataframe if there is no header.
I would do something more like

```
headers = getattr(data_frame, {})
```

at the very beginning, then check everything on this object (no need for the `_has_headers()` function). Add TYPE to it here in the last check and then set `data_frame.headers = headers` at the end

Reply:
> This is not good. Not a valid dataframe if there is no header.

We've been accepting it so far..

Also the validation function does not return anything. So if there are no headers, then we're adding "TYPE" to some dict which goes nowhere and we haven't helped the user.